causalpy/exp_inverse_propensity_weighting.py

```diff
@@ -0,0 +1,433 @@
+        outcome_ntrt = X[t == 0][self.outcome_variable]
+        i_propensity0 = X[t == 0]["i_ps"]
+        i_propensity1 = X[t == 1]["i_ps"]
+        weighted_outcome1 = t[t == 1] * outcome_trt * i_propensity1
+        weighted_outcome0 = (1 - t[t == 0]) * outcome_ntrt * i_propensity0
+        return weighted_outcome0, weighted_outcome1, n_ntrt, n_trt
+
+    def make_doubly_robust_adjustment(self, ps):
+        """The doubly robust weighting scheme is also
+        discussed in Aronow and Miller, but a bit more generally
+        than our implementation here. Here we have specified
+        the outcome model to be a simple OLS model.
+        In this way the compromise between the outcome model and
+        the propensity model is always done with OLS."""
+        X = pd.DataFrame(self.X, columns=self.labels)
+        X["ps"] = ps
+        t = self.t.flatten()
+        m0 = sk_lin_reg().fit(X[t == 0].astype(float), self.y[t == 0])
+        m1 = sk_lin_reg().fit(X[t == 1].astype(float), self.y[t == 1])
+        m0_pred = m0.predict(X)
+        m1_pred = m1.predict(X)
+        ## Compromise between outcome and treatement assignment model
+        weighted_outcome0 = (1 - t) * (self.y - m0_pred) / (1 - X["ps"]) + m0_pred
+        weighted_outcome1 = t * (self.y - m1_pred) / X["ps"] + m1_pred
+        return weighted_outcome0, weighted_outcome1, None, None
+
+    def get_ate(self, i, idata, method="doubly_robust"):
+        ### Post processing the sample posterior distribution for propensity scores
+        ### One sample at a time.
+        ps = idata["posterior"]["p"].stack(z=("chain", "draw"))[:, i].values
+        if method == "robust":
+            (
+                weighted_outcome_ntrt,
+                weighted_outcome_trt,
+                n_ntrt,
+                n_trt,
+            ) = self.make_robust_adjustments(ps)
+            ntrt = weighted_outcome_ntrt.sum() / n_ntrt
+            trt = weighted_outcome_trt.sum() / n_trt
+        elif method == "raw":
+            (
+                weighted_outcome_ntrt,
+                weighted_outcome_trt,
+                n_ntrt,
+                n_trt,
+            ) = self.make_raw_adjustments(ps)
+            ntrt = weighted_outcome_ntrt.sum() / n_ntrt
+            trt = weighted_outcome_trt.sum() / n_trt
+        elif method == "overlap":
+            (
+                weighted_outcome_ntrt,
+                weighted_outcome_trt,
+                n_ntrt,
+                n_trt,
+            ) = self.make_overlap_adjustments(ps)
+            ntrt = np.sum(weighted_outcome_ntrt) / np.sum(n_ntrt)
+            trt = np.sum(weighted_outcome_trt) / np.sum(n_trt)
+        else:
+            (
+                weighted_outcome_ntrt,
+                weighted_outcome_trt,
+                n_ntrt,
+                n_trt,
+            ) = self.make_doubly_robust_adjustment(ps)
+            trt = np.mean(weighted_outcome_trt)
+            ntrt = np.mean(weighted_outcome_ntrt)
```

Comment:
Each of these conditions should be individual functions and have their own unit-test. Then `get_ate` simply can call them.

Reply:
I've created a separate issue about this https://github.com/pymc-labs/CausalPy/issues/389